includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+<?php
+
+namespace Automattic\WP\Cron_Control;
+
+class Events_Store extends Singleton {
+	/**
+	 * PLUGIN SETUP
+	 */
+
+	/**
+	 * Class properties
+	 */
+	const TABLE_SUFFIX = 'a8c_cron_control_jobs';
+
+	const DB_VERSION        = 1;
+	const DB_VERSION_OPTION = 'a8c_cron_control_db_version';
+	const TABLE_CREATE_LOCK = 'a8c_cron_control_creating_table';
+
+	const STATUS_PENDING   = 'pending';
+	const STATUS_RUNNING   = 'running';
+	const STATUS_COMPLETED = 'complete';
+
+	const CACHE_KEY = 'a8c_cron_ctrl_option';
+
+	private $job_creation_suspended = false;
+
+	/**
+	 * Register hooks
+	 */
+	protected function class_init() {
+		// Check that the table exists and is the correct version
+		$this->prepare_tables();
+
+		// Option interception
+		add_filter( 'pre_option_cron', array( $this, 'get_option' ) );
+		add_filter( 'pre_update_option_cron', array( $this, 'update_option' ), 10, 2 );
+
+		// Disallow duplicates
+		add_filter( 'schedule_event', array( $this, 'block_creation_if_job_exists' ) );
+	}
+
+	/**
+	 * Build appropriate table name for this install
+	 */
+	public function get_table_name() {
+		global $wpdb;
+
+		return $wpdb->prefix . self::TABLE_SUFFIX;
+	}
+
+	/**
+	 * Build array of valid event statuses
+	 */
+	public function get_allowed_statuses() {
+		return array(
+			self::STATUS_PENDING,
+			self::STATUS_RUNNING,
+			self::STATUS_COMPLETED,
+		);
+	}
+
+	/**
+	 * Create the plugin's DB table when necessary
+	 */
+	protected function prepare_tables() {
```

Comment:
Initially I did, but until the table is created, every request tries to schedule the events a site defined. The `wp_cache_add()` is meant to guard against this. Instead, or perhaps in addition, the option should be set sooner, to cut down on how many processes can create the table.